### PR TITLE
[Core] Add workaround for cross-tenant authentication with Track 2 SDKs

### LINKED ISSUE #16691: `az vnet peering create` no longer works cross-tenant
## Describe the bug

**Command Name**
`az network vnet peering create`

**Errors:**
```
ForbiddenError: (LinkedAuthorizationFailed) The client has permission to perform action 'Microsoft.Network/virtualNetworks/peer/action' on scope '/subscriptions/$SECONDARY_SUBSCRIPTION/resourceGroups/mzink-test/providers/Microsoft.Network/virtualNetworks/mzink-test-vnet/virtualNetworkPeerings/mzink-test-vnet-to-mzink-vnet', however the current tenant '$SECONDARY_TENANT' is not authorized to access linked subscription '$PRIMARY_SUBSCRIPTION'.
```

## To Reproduce:

- Give a user in `$SECONDARY_TENANT` the `Network Contributor` role on a Virtual Network (`$PRIMARY_VNET_ID`) that lives in `$PRIMARY_SUBSCRIPTION` within `$PRIMARY_TENANT`.
- Log in to both tenants as the user from the secondary tenant:
  - `az login -t $PRIMARY_TENANT`
  - `az login -t $SECONDARY_TENANT`
- `az network vnet peering create --allow-vnet-access --allow-forwarded-traffic --allow-gateway-transit -n secondary-to-primary --remote-vnet $PRIMARY_VNET_ID -g $SECONDARY_VNET_RG --vnet-name $SECONDARY_VNET_NAME --subscription $SECONDARY_SUBSCRIPTION`

## Expected Behavior

The VNet peering should be created. This was the case when using v2.12 of the CLI.

## Environment Summary
```
Linux-5.4.0-58-generic-x86_64-with-debian-buster-sid
Python 3.6.10
Installer: DEB

azure-cli 2.18.0

```
## Additional Context

The same also happens when attempting the reverse direction (primary VNet -> secondary VNet peering), with the error output showing the inverse IDs.

We have also developed on v2.17 of the CLI, which produces this same error, and v2.16, which instead produced a `BadRequestError`. Between all four versions (including v2.12 which does not error and successfully creates the peering) the only differences I could see when running with `--verbose --debug` were:

1. the ordering of the JSON body fields
2. the API version (2020-06-01 for v2.12, then 2020-07-01 for v2.16+)

I tried testing all combinations (old API version & old ordering, new API version & old ordering, etc.) on both v2.12 and v2.17 using the `az rest` command, and got the new `LinkedAuthorizationFailed` error in all cases.

### PR DESCRIPTION
**Pull request by jiasli**
Fix #16691: `az vnet peering create` no longer works cross-tenant

⚠ This is only a temporary workaround for https://github.com/Azure/azure-sdk-for-python/issues/8313.

## Changes

Manually add external tokens to `x-ms-authorization-auxiliary` header.

### Regarding auto-refresh

As far as I know and also tested for [asynchronous Azure operations](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations) (long-running operations)

```
az network vnet peering create 
    --name myVnetAToMyVnetB
    --resource-group myResourceGroupA
    --vnet-name myVnetA 
    --remote-vnet /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/myResourceGroupB/providers/Microsoft.Network/VirtualNetworks/myVnetB 
    --allow-vnet-access 
    --debug
```

`x-ms-authorization-auxiliary` is only necessary in the first request:

```
azure.core.pipeline.policies._universal: Request URL: 'https://management.azure.com/subscriptions/414af076-009b-4282-9a0a-acf75bcb037e/resourceGroups/myResourceGroupA/providers/Microsoft.Network/virtualNetworks/myVnetA/virtualNetworkPeerings/myVnetAToMyVnetB?api-version=2020-07-01'
azure.core.pipeline.policies._universal: Request method: 'PUT'
azure.core.pipeline.policies._universal: Request headers:
azure.core.pipeline.policies._universal:     'x-ms-authorization-auxiliary': 'Bearer **********'
azure.core.pipeline.policies._universal:     'Authorization': '*****'
azure.core.pipeline.policies._universal: Request body:
azure.core.pipeline.policies._universal: ... "remoteVirtualNetwork": {"id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/myResourceGroupB/providers/Microsoft.Network/VirtualNetworks/myVnetB"} ...

azure.core.pipeline.policies._universal: Response status: 200
azure.core.pipeline.policies._universal: Response headers:
azure.core.pipeline.policies._universal:     'Azure-AsyncOperation': 'https://management.azure.com/subscriptions/414af076-009b-4282-9a0a-acf75bcb037e/providers/Microsoft.Network/locations/eastus/operations/78792a4b-d45f-4037-8149-18baa4fbff5e?api-version=2020-07-01'
```

<sub> The response is a `200` with `Azure-AsyncOperation`, which contradicts [Track asynchronous Azure operations](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations). (https://github.com/Azure/azure-rest-api-specs/issues/12828) </sub>

The following pooling request won't need `x-ms-authorization-auxiliary` and can also succeed:

```http
GET https://management.azure.com/subscriptions/<subId>/providers/Microsoft.Network/locations/eastus/operations/7de4e22f-7876-4596-9857-bcee50d703da?api-version=2020-07-01
Authorization: Bearer *****

200
{"status": "Succeeded"}
```

Therefore, we can believe tokens in `x-ms-authorization-auxiliary` is only used once, so auto-refresh within a long-running operation won't be necessary.

## Alternative solutions

Develop a custom policy `ExternalBearerTokenCredentialPolicy` and add it to the client. However, adding a custom policy to a client is current impracticable (https://github.com/Azure/azure-sdk-for-python/issues/16519).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py [hunk 1]
```diff
@@ -194,6 +194,19 @@ def _get_mgmt_service_client(cli_ctx,
...
+        # As a temporary workaround, manually add external tokens to 'x-ms-authorization-auxiliary' header.
+        #   https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant
+        if aux_subscriptions or aux_tenants:
+            _, _, _, external_tenant_tokens = cred.get_all_tokens(*resource_to_scopes(resource))
+
+            # external_tenant_tokens can be [] if no external tenant is involved.
+            if external_tenant_tokens:
+                # Hard-code scheme to 'Bearer' as _BearerTokenCredentialPolicyBase._update_headers does.
+                client_kwargs['headers']['x-ms-authorization-auxiliary'] = \
+                    ', '.join("Bearer {}".format(t[1]) for t in external_tenant_tokens)
```
Comment (evelyn-ys) [inline src/azure-cli-core/azure/cli/core/commands/client_factory.py:208]: Can both `,` and `;` be used as the delimiter for each token?

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:208]: I simply don't know. We need to do some test. 

The exiting code 

https://github.com/Azure/azure-cli/blob/d07ee662c8cb49978901fc4dc11ddad46c89c27a/src/azure-cli-core/azure/cli/core/adal_authentication.py#L142

contradicts the document [Authenticate requests across tenants](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant)

```
Bearer <auxiliary-token1>, EncryptedBearer <auxiliary-token2>, Bearer <auxiliary-token3>
```

That's why this PR is still marked as a **draft**. 😉

**Comment by jiasli** [comment]
@jsntcy @msyyc, please kindly test for Network module and add some `LiveScenarioTest`s.

I have already written some testing script: https://github.com/Azure/azure-cli/issues/16691#issuecomment-770753505

You may refer to `VMCrossTenantUpdateScenarioTest` to write similar tests:

https://github.com/Azure/azure-cli/blob/fa7939bf41a37ebf5892fe88774675b91e7676c6/src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py#L4877-L4913

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 1 above)
Comment (zhoxing-ms) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py]: According to my experiment of ARM deployment, the correct delimiter should be `, `.
When the delimiter is `;`, the service will return the following error:
```
{"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}}
```

Comment (zhoxing-ms) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py]: According to my experiment of ARM deployment, the correct delimiter should be `, `.
When the delimiter is `;`, the service returns the following error:
```
{"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}}
```

**Review by zhoxing-ms — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 1 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py]: True! Also verified with:

```
> az rest --url "https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourcegroups?api-version=2020-10-01" --headers "x-ms-authorization-auxiliary=Bearer token1; Bearer token2" --verbose
Request URL: 'https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourcegroups?api-version=2020-10-01'
Request method: 'GET'
Request headers:
    'User-Agent': 'python/3.8.8 (Windows-10-10.0.19041-SP0) AZURECLI/2.20.0 (PIP)'
    'Accept-Encoding': 'gzip, deflate'
    'Accept': '*/*'
    'Connection': 'keep-alive'
    'x-ms-authorization-auxiliary': 'Bearer token1; Bearer token2'
    'x-ms-client-request-id': 'aad4a80f-5a67-425c-bc3d-9b5a2f471959'
    'CommandName': 'rest'
    'ParameterSetName': '--url --headers --verbose'
    'Authorization': 'Bearer eyJ0eXAiOiJKV...'
Request body:
None
Response status: 401
Response headers:
    'Cache-Control': 'no-cache'
    'Pragma': 'no-cache'
    'Content-Type': 'application/json; charset=utf-8'
    'Expires': '-1'
    'WWW-Authenticate': 'Bearer authorization_uri="https://login.windows.net/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a", error="invalid_token", error_description="Invalid auxiliary tokens \'Bearer token1; Bearer token2\'."'
    'x-ms-failure-cause': 'gateway'
    'x-ms-request-id': '08deb134-9056-4afa-a468-9025f17246f3'
    'x-ms-correlation-request-id': '08deb134-9056-4afa-a468-9025f17246f3'
    'x-ms-routing-request-id': 'SOUTHEASTASIA:20210311T120202Z:08deb134-9056-4afa-a468-9025f17246f3'
    'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
    'X-Content-Type-Options': 'nosniff'
    'Date': 'Thu, 11 Mar 2021 12:02:01 GMT'
    'Connection': 'close'
    'Content-Length': '272'
Response content:
{"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}}
Unauthorized({"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}})
```

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py]: True! Also verified with:

```
> az rest --url "https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourcegroups?api-version=2020-10-01" --headers "x-ms-authorization-auxiliary=Bearer token1; Bearer token2" --verbose
Request URL: 'https://management.azure.com/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourcegroups?api-version=2020-10-01'
Request method: 'GET'
Request headers:
    'Authorization': 'Bearer eyJ0eXAiOiJKV...'
    'x-ms-authorization-auxiliary': 'Bearer token1; Bearer token2'    
    ...

Response status: 401
Response headers:
    'WWW-Authenticate': 'Bearer authorization_uri="https://login.windows.net/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a", error="invalid_token", error_description="Invalid auxiliary tokens \'Bearer token1; Bearer token2\'."'
    ...
Response content:
{"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}}
Unauthorized({"error":{"code":"InvalidAuxiliaryTokens","message":"Authentication failed for auxiliary token: 'The auxiliary tokens should have proper format \"'scheme' 'value'\". The '1' auxiliary token(s) were not found in proper format. Invalid tokens are in the response header.'"}})
```

This means the doc is right and the current CLI code is wrong.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 2]
```diff
@@ -3282,31 +3282,13 @@ def create_image_version(cmd, resource_group_name, gallery_name, gallery_image_n
...
     client = _compute_client_factory(cmd.cli_ctx, aux_subscriptions=aux_subscriptions)
 
-    # Auxiliary tokens, pass it to init or operation
-    external_bearer_token = None
-    if aux_subscriptions:
-        profile = Profile(cli_ctx=cmd.cli_ctx)
-        resource = cmd.cli_ctx.cloud.endpoints.active_directory_resource_id
-        cred, _, _ = profile.get_login_credentials(resource=resource,
-                                                   aux_subscriptions=aux_subscriptions)
-        _, _, _, external_tokens = cred.get_all_tokens('https://management.azure.com/.default')
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3300]: By moving this to core, it now works for sovereign cloud (solves https://github.com/Azure/azure-cli/pull/15750#discussion_r546513080).

**Review by msyyc — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 3]
```diff
@@ -573,6 +574,10 @@ def get_login_credentials(self, resource=None, subscription_id=None, aux_subscri
                 if sub[_TENANT_ID] != account[_TENANT_ID]:
                     external_tenants_info.append(sub[_TENANT_ID])
 
+        if external_tenants_info and (identity_type or in_cloud_console()):
+            raise CLIError("Cross-tenant authentication is not supported by managed identity and Cloud Shell. "
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/_profile.py:578]: CLIError is deprecated.

File: src/azure-cli-core/azure/cli/core/adal_authentication.py [hunk 4]
```diff
@@ -35,7 +35,7 @@ def _get_token(self, sdk_resource=None):
         """
         external_tenant_tokens = None
         try:
-            scheme, token, full_token = self._token_retriever(sdk_resource)
+            scheme, token, token_entry = self._token_retriever(sdk_resource)
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/adal_authentication.py:38]: Is it just a rename operation? Does the object structure change?

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/adal_authentication.py:38]: Yes. No.

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py [hunk 5]
```diff
@@ -194,6 +194,16 @@ def _get_mgmt_service_client(cli_ctx,
         client_kwargs.update(_prepare_client_kwargs_track2(cli_ctx))
         client_kwargs['credential_scopes'] = resource_to_scopes(resource)
 
+        # Track 2 currently lacks the ability to take external credentials.
+        #   https://github.com/Azure/azure-sdk-for-python/issues/8313
+        # As a temporary workaround, manually add external tokens to 'x-ms-authorization-auxiliary' header.
+        #   https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/authenticate-multi-tenant
+        if getattr(cred, "_external_tenant_token_retriever", None):
+            _, _, _, external_tenant_tokens = cred.get_all_tokens(*resource_to_scopes(resource))
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/commands/client_factory.py:202]: What is `*`

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py [hunk 6]
```diff
@@ -194,6 +194,16 @@ def _get_mgmt_service_client(cli_ctx,
...
+            # Hard-code scheme to 'Bearer' as _BearerTokenCredentialPolicyBase._update_headers does.
+            client_kwargs['headers']['x-ms-authorization-auxiliary'] = \
+                ', '.join("Bearer {}".format(t[1]) for t in external_tenant_tokens)
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/commands/client_factory.py:205]: Is there a pipeline support?

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 5 above)
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/commands/client_factory.py:202]:
```suggestion
            *_, external_tenant_tokens = cred.get_all_tokens(*resource_to_scopes(resource))
```

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 6 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:205]: I will make a new policy and add it to the pipeline in the future, utilizing https://github.com/Azure/azure-sdk-for-python/pull/17340.

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 5 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:202]: Readability is not so good.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 7]
```diff
@@ -3363,8 +3345,7 @@ def create_image_version(cmd, resource_group_name, gallery_name, gallery_image_n
         gallery_name=gallery_name,
         gallery_image_name=gallery_image_name,
         gallery_image_version_name=gallery_image_version,
-        gallery_image_version=image_version,
-        headers={'x-ms-authorization-auxiliary': external_bearer_token}
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:3367]: You deleted lots of code in vm/custom.py and didn't add anything. How are external tokens handled now? Is another PR required to fix it?

File: src/azure-cli-core/azure/cli/core/commands/client_factory.py (hunk 5 above)
Comment (qwordy) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:202]: It is an advanced feature of Python.
https://www.python.org/dev/peps/pep-3132/

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/client_factory.py:202]: See [More on Defining Functions](https://docs.python.org/3.9/tutorial/controlflow.html#more-on-defining-functions).

File: src/azure-cli-core/azure/cli/core/_profile.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/_profile.py:578]: This error is difficult to classify.

File: src/azure-cli/azure/cli/command_modules/vm/custom.py (hunk 7 above)
Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/vm/custom.py:3367]: > You deleted lots of code in vm/custom.py and didn't add anything.

The logic is not deleted, but moved to `_get_mgmt_service_client`. 

> How are external tokens handled now? 

External tokens are handled automatically when `aux_subscriptions` is provided to `_get_mgmt_service_client`.

```py
client = _compute_client_factory(cmd.cli_ctx, aux_subscriptions=aux_subscriptions)
```

Is another PR required to fix it?

> It already works. No other PR needed.

**Comment by qwordy** [comment]
Have you run live test for `sig` commands?

**Comment by jiasli** [comment]
> Have you run live test for `sig` commands?

No. You may take some time to run it.

**Comment by qwordy** [comment]
Testing results of VM. Fail. Subscription is not found. Investigating.
https://clitestresultstac.blob.core.windows.net/20210406-124636-446567/vm.report.parallel.html

**Comment by jiasli** [comment]
> Testing results of VM. Fail. Subscription is not found. Investigating.

It is because the test relies on Microsoft tenant 

https://github.com/Azure/azure-cli/blob/ba16edd26870f170a4f7a1f9d63aaca5b8c00bbf/src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py#L3846

but your test account doesn't have permission there.

**Review by qwordy — APPROVED** [review]

**Pull request merged**